schema-engine/core/src/url.rs

```diff
@@ -0,0 +1,249 @@
+    url: String,
+    shadow_database_url: Option<String>,
+}
+
+/// Errors produced while validating datasource URLs supplied from configuration.
+#[derive(thiserror::Error, Debug)]
+pub enum DatasourceError {
+    #[error("`datasource.{_0}` in `prisma.config.ts` must not be an empty string.")]
+    Empty(&'static str),
+
+    #[error(
+        "`datasource.{_0}` in `prisma.config.ts` must be a direct URL that points directly to the database. Using `prisma:` in the URL scheme is not allowed."
+    )]
+    AccelerateUrl(&'static str),
+}
+
+impl From<&DatasourceError> for CoreError {
+    fn from(error: &DatasourceError) -> Self {
+        CoreError::new_invalid_datasource_error(error)
+    }
+}
+
+impl From<DatasourceError> for CoreError {
+    fn from(error: DatasourceError) -> Self {
+        Self::from(&error)
+    }
+}
+
+impl TryFrom<DatasourceUrls> for ValidatedDatasourceUrls {
+    type Error = DatasourceError;
+
+    fn try_from(urls: DatasourceUrls) -> Result<Self, Self::Error> {
+        validate_datasource_url(&urls.url, "url")?;
+
+        if let Some(shadow_database_url) = &urls.shadow_database_url {
+            validate_datasource_url(shadow_database_url, "shadowDatabaseUrl")?;
+        }
+
+        Ok(ValidatedDatasourceUrls {
+            url: urls.url,
+            shadow_database_url: urls.shadow_database_url,
+        })
+    }
+}
+
+fn validate_datasource_url(url: &str, name: &'static str) -> Result<(), DatasourceError> {
+    if url.is_empty() {
+        return Err(DatasourceError::Empty(name));
+    }
+
+    if url.starts_with("prisma://") {
+        return Err(DatasourceError::AccelerateUrl(name));
+    }
+
+    Ok(())
+}
+
+impl ValidatedDatasourceUrls {
+    /// Returns the validated primary datasource URL.
+    pub fn url(&self) -> &str {
+        &self.url
+    }
+
+    /// Returns the validated shadow database URL, if any.
+    pub fn shadow_database_url(&self) -> Option<&str> {
+        self.shadow_database_url.as_deref()
+    }
+
+    /// Resolves relative paths in the URL against the provided configuration directory.
+    pub fn url_with_config_dir(&self, flavour: Flavour, config_dir: &Path) -> Cow<'_, str> {
+        set_config_dir(flavour, config_dir, &self.url)
+    }
+
+    /// Returns both URLs with relative file paths rewritten relative to the configuration directory.
+    pub fn with_config_dir(&self, flavour: Flavour, config_dir: &Path) -> DatasourceUrlsWithConfigDir<'_> {
+        DatasourceUrlsWithConfigDir {
+            url: set_config_dir(flavour, config_dir, &self.url),
+            shadow_database_url: self
+                .shadow_database_url
+                .as_deref()
+                .map(|url| set_config_dir(flavour, config_dir, url)),
+        }
+    }
+}
+
+/// Datasource URLs with relative paths resolved against the configuration directory.
+#[derive(Debug, Clone)]
+pub struct DatasourceUrlsWithConfigDir<'a> {
+    url: Cow<'a, str>,
+    shadow_database_url: Option<Cow<'a, str>>,
+}
```

Comment:
Previously we didn't resolve `shadowDatabaseUrl` against the config dir, so the behaviour was wrong if, e.g., you needed to use client certificates for the shadow db. Using proper static types fixes bugs again.